### PR TITLE
Uplift drop-bounds lint from clippy

### PR DESCRIPTION
**Pull request by notriddle**
Bounds on `T: Drop` do nothing, so they should warn.

### CODE REVIEW

File: src/librustc_lint/traits.rs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+            let def_id = trait_predicate.trait_ref.def_id;
+            if cx.tcx.is_diagnostic_item(sym::Drop, def_id) {
+                // Explicitly allow `impl Drop`, a drop-guards-as-Voldemort-type pattern.
+                if trait_predicate.trait_ref.self_ty().is_impl_trait() {
+                    continue;
+                }
+                cx.struct_span_lint(DROP_BOUNDS, span, |lint| {
+                    let msg = format!(
+                        "bounds on {} are useless, consider instead \
+                            using `{}` to detect if a type has a destructor",
```
Comment (lzutao) [inline src/librustc_lint/traits.rs:43]:
```suggestion
                        "bounds on `{}` are useless, consider instead \
                         using `{}` to detect if a type has a destructor",
```

File: library/core/src/ops/drop.rs [hunk 2]
```diff
@@ -136,6 +136,7 @@
 /// [`Copy`]: ../../std/marker/trait.Copy.html
 #[lang = "drop"]
 #[stable(feature = "rust1", since = "1.0.0")]
+#[rustc_diagnostic_item = "Drop"]
```
Comment (petrochenkov) [inline library/core/src/ops/drop.rs:139]: `Drop` is a lang item, which already makes it known to the compiler, so it doesn't need to be a diagnostic item as well.

File: src/test/ui/drop-bounds/drop-bounds.rs [hunk 3]
```diff
@@ -0,0 +1,18 @@
+#![deny(drop_bounds)]
+#![allow(unused)]
```
Comment (petrochenkov) [inline src/test/ui/drop-bounds/drop-bounds.rs:2]: UI tests are `allow(unused)` by default, so this attribute is not necessary.

File: src/librustc_lint/traits.rs [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        use rustc_middle::ty::PredicateAtom::*;
+
+        let def_id = cx.tcx.hir().local_def_id(item.hir_id);
+        let predicates = cx.tcx.explicit_predicates_of(def_id);
+        for &(predicate, span) in predicates.predicates {
+            let trait_predicate = match predicate.skip_binders() {
+                Trait(trait_predicate, _constness) => trait_predicate,
+                _ => continue,
+            };
+            let needs_drop = match cx.tcx.get_diagnostic_item(sym::needs_drop) {
```
Comment (petrochenkov) [inline src/librustc_lint/traits.rs:30]: This can be moved closer to the `needs_drop`'s point of use to avoid extra work in the common case.

**Comment by petrochenkov** [comment]
This looks reasonable to me.
Could you open an MCP for this? (Example: https://github.com/rust-lang/compiler-team/issues/346)

File: src/test/ui/drop-bounds/drop-bounds.rs [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+{
+}
+fn baz(_x: impl Drop) {} //~ ERROR
+struct Foo<T: Drop> { //~ ERROR
+  _x: T
+}
+struct Bar<U> where U: Drop { //~ ERROR
+  _x: U
+}
+trait Baz: Drop { //~ ERROR
```
Comment (Aaron1011) [inline src/test/ui/drop-bounds/drop-bounds.rs:16]: Can you add a test for an `impl` block that uses a `Drop` bound?

**Comment by notriddle** [comment]
https://github.com/rust-lang/compiler-team/issues/347 Is the MCP

File: src/librustc_lint/traits.rs [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+            let def_id = trait_predicate.trait_ref.def_id;
+            if cx.tcx.is_diagnostic_item(sym::Drop, def_id) {
+                // Explicitly allow `impl Drop`, a drop-guards-as-Voldemort-type pattern.
+                if trait_predicate.trait_ref.self_ty().is_impl_trait() {
+                    continue;
+                }
+                cx.struct_span_lint(DROP_BOUNDS, span, |lint| {
+                    let msg = format!(
+                        "bounds on `{}` are useless, consider instead \
+                            using `{}` to detect if a type has a destructor",
```
Comment (lzutao) [inline src/librustc_lint/traits.rs:43]:
```suggestion
                        "bounds on `{}` are useless, consider instead \
                         using `{}` to detect if a type has a destructor",
```

File: library/core/src/ops/drop.rs (hunk 2 above)
Comment (notriddle) [reply library/core/src/ops/drop.rs]: That's fixed.

**Comment by petrochenkov** [comment]
r=me on the implementation after squashing commits.
Looks like this is still waiting on team though.

**Comment by flip1995** [comment]
Should we directly deprecate the lint from Clippy in this PR or do it after this PR is merged in Clippy and sync it back to rustc?

I can guide through the process of deprecating Clippy lints.

**Comment by spastorino** [comment]
This is no longer waiting on team, MCP has been accepted.

**Comment by petrochenkov** [comment]
r=me after squashing commits.

**Comment by notriddle** [comment]
This PR predates https://github.com/rust-lang/rust/pull/76549, which added a requirement that lints have docstrings. Oops.

**Comment by petrochenkov** [comment]
Some clippy tests need to be updated.

**Comment by flip1995** [comment]
Either we update the Clippy tests here, then sync these changes to Clippy, then deprecate the lint in Clippy, and then sync it back to rustc

or

We deprecate the lint here, don't have to go through the whole sync process twice and make Clippy test-pass at the same time.

---

Instructions how to deprecate a Clippy lint:
1. ~~Remove this file: https://github.com/rust-lang/rust/blob/4dedf5edd51d0e0b1b45a61403842f8406e13b2c/src/tools/clippy/clippy_lints/src/drop_bounds.rs#L1~~
2. ~~Add this lint to this file https://github.com/rust-lang/rust/blob/4dedf5edd51d0e0b1b45a61403842f8406e13b2c/src/tools/clippy/clippy_lints/src/deprecated_lints.rs#L1~~
3. ~~Remove the test file https://github.com/rust-lang/rust/blob/4dedf5edd51d0e0b1b45a61403842f8406e13b2c/src/tools/clippy/tests/ui/drop_bounds.rs#L1~~
4. 
    a. ~~temporarily add `src/tools/clippy/clippy_dev` to `workspace.members` in the `Cargo.toml`~~
    b. ~~`cd` into `src/tools/clippy` and run `cargo dev update_lints`~~
5. ~~Clean up Clippy until `./x.py check src/tools/clippy` passes~~
6. ~~Add a deprecation test and update the `deprecated.stderr` reference file.~~

Or just apply this patch: https://gist.github.com/flip1995/bbde96b04ca768f4ed7713dd53f23641

**Review by flip1995 — APPROVED** [review]
ACK Clippy change 👍

**Pull request merged**